cleverhans/future/tf2/attacks/spsa.py

```diff
@@ -0,0 +1,427 @@
+
+  Args:
+    :param loss_fn: A callable which takes `input_image` and `label` as
+                    arguments, and returns a batch of loss values. Same
+                    interface as TensorOptimizer.
+    :param input_image: Tensor, a batch of images
+    :param label: Tensor, a batch of labels
+    :param epsilon: float, the L-infinity norm of the maximum allowable
+                    perturbation
+    :param num_steps: int, the number of steps of gradient descent
+    :param clip_min: float, minimum pixel value
+    :param clip_max: float, maximum pixel value
+    :param optimizer: A `TensorOptimizer` object
+    :param project_perturbation: A function, which will be used to enforce
+                                 some constraint. It should have the same
+                                 signature as `_project_perturbation`.
+    :param early_stop_loss_threshold: A float or None. If specified, the attack will end if the loss is below
+       `early_stop_loss_threshold`.
+        Enabling this option can have several different effects:
+          - Setting the threshold to 0. guarantees that if a successful attack is found, it is returned.
+            This increases the attack success rate, because without early stopping the optimizer can accidentally
+            bounce back to a point where the attack fails.
+          - Early stopping can make the attack run faster because it may run for fewer steps.
+          - Early stopping can make the attack run slower because the loss must be calculated at each step.
+            The loss is not calculated as part of the normal SPSA optimization procedure.
+            For most reasonable choices of hyperparameters, early stopping makes the attack much faster because
+            it decreases the number of steps dramatically.
+    :param is_debug: A bool. If True, print debug info for attack progress.
+
+  Returns:
+    adversarial version of `input_image`, with L-infinity difference less than epsilon, which tries
+    to minimize loss_fn.
+
+  Note that this function is not intended as an Attack by itself. Rather, it is designed as a helper
+  function which you can use to write your own attack methods. The method uses a tf.while_loop to
+  optimize a loss function in a single sess.run() call.
+  """
+  assert num_steps is not None
+  if is_debug:
+    with tf.device("/cpu:0"):
+      tf.print("Starting PGD attack with epsilon: %s" % epsilon)
+
+  init_perturbation = tf.random.uniform(tf.shape(input_image),
+                                        minval=tf.cast(-epsilon, input_image.dtype),
+                                        maxval=tf.cast(epsilon, input_image.dtype),
+                                        dtype=input_image.dtype)
+  init_perturbation = project_perturbation(init_perturbation, epsilon, input_image,
+                                           clip_min=clip_min, clip_max=clip_max)
+  init_optim_state = optimizer.init_state([init_perturbation])
+  nest = tf.nest
```

Comment:
It would be easier to read the code if all instances of `nest` are replaced with `tf.nest`

Reply:
(now that `nest` is no longer in `tf.contrib`)